src/layers.c

```diff
@@ -118,45 +118,116 @@ handle_output_destroy(struct wl_listener *listener, void *data)
 	wlr_layer_surface_v1_destroy(layer->scene_layer_surface->layer_surface);
 }
 
-static void
-process_keyboard_interactivity(struct lab_layer_surface *layer)
+static inline bool
+has_exclusive_interactivity(struct wlr_scene_layer_surface_v1 *scene)
 {
-	struct wlr_layer_surface_v1 *layer_surface = layer->scene_layer_surface->layer_surface;
-	struct seat *seat = &layer->server->seat;
+	return scene->layer_surface->current.keyboard_interactive
+		== ZWLR_LAYER_SURFACE_V1_KEYBOARD_INTERACTIVITY_EXCLUSIVE;
+}
 
-	if (layer_surface->current.keyboard_interactive
-			&& layer_surface->current.layer >= ZWLR_LAYER_SHELL_V1_LAYER_TOP) {
+/*
+ * Try to transfer focus other layer-shell clients with exclusive focus on the
+ * output nearest to the cursor. In none exist (which is likely to generally be
+ * the case) just unset layer focus and try to give it to the topmost toplevel
+ * if one exists.
+ */
+static void
+try_to_focus_next_layer_or_toplevel(struct server *server)
+{
+	struct seat *seat = &server->seat;
+	struct output *output = output_nearest_to_cursor(server);
+
+	int overlay = ZWLR_LAYER_SHELL_V1_LAYER_OVERLAY;
+	int top = ZWLR_LAYER_SHELL_V1_LAYER_TOP;
+	for (int i = overlay; i >= top; i--) {
+		struct wlr_scene_tree *tree = output->layer_tree[i];
+		struct wlr_scene_node *node;
 		/*
-		 * Give keyboard focus to surface if
-		 * - keyboard-interactivity is 'exclusive' or 'on-demand'; and
-		 * - surface is in top/overlay layers; and
-		 * - currently focused layer has a lower precedence
-		 *
-		 * In other words, when dealing with two surfaces with
-		 * exclusive/on-demand keyboard-interactivity (firstly the
-		 * currently focused 'focused_layer' and secondly the
-		 * 'layer_surface' for which we're just responding to a
-		 * map/commit event), the following logic applies:
-		 *
-		 * | focused_layer | layer_surface | who gets keyboard focus |
-		 * |---------------|---------------|-------------------------|
-		 * | overlay       | top           | focused_layer           |
-		 * | overlay       | overlay       | layer_surface           |
-		 * | top           | top           | layer_surface           |
-		 * | top           | overlay       | layer_surface           |
+		 * In wlr_scene.c they were added at end of list so we
+		 * iterate in reverse to process last client first.
 		 */
+		wl_list_for_each_reverse(node, &tree->children, link) {
+			struct lab_layer_surface *layer = node_layer_surface_from_node(node);
+			struct wlr_scene_layer_surface_v1 *scene = layer->scene_layer_surface;
+			struct wlr_layer_surface_v1 *layer_surface = scene->layer_surface;
```

Comment:
Is it ensured that there can be no other nodes in that tree and that all `lab_layer_surfaces` do in fact have a `scene_layer_surface` at all times and that in turn also has a `layer_surface`? Otherwise we may want to check for that first and `continue`.

Reply:
I'm fairly comfortable (if there is ever such a thing!)

> Is it ensured that there can be no other nodes in that tree

I think so. And in `node_layer_surface_from_node()` we assert() node-descriptor type.

> and that all `lab_layer_surfaces` do in fact have a `scene_layer_surface` at all times and that in turn also has a `layer_surface`? Otherwise we may want to check for that first and `continue`.

This is consistent with the rest of `src/layers.c`. The `wlr_layer_surface_v1` is the primary object. The other two are just wrapper structs.